runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java

```diff
@@ -104,29 +104,24 @@ public Layout getCurrentLayout() {
                 if (re.getCause() instanceof TimeoutException) {
                     log.warn("Timeout executing remote call, invalidating view and retrying in {}s",
                             retryRate);
-                    runtime.invalidateLayout();
-                    Sleep.sleepUninterruptibly(retryRate);
                 } else if (re instanceof ServerNotReadyException) {
                     log.warn("Server still not ready. Waiting for server to start "
                             + "accepting requests.");
-                    Sleep.sleepUninterruptibly(retryRate);
                 } else if (re instanceof WrongEpochException) {
                     WrongEpochException we = (WrongEpochException) re;
                     log.warn("Got a wrong epoch exception, updating epoch to {} and "
                             + "invalidate view", we.getCorrectEpoch());
-                    runtime.invalidateLayout();
                 } else if (re instanceof NetworkException) {
                     log.warn("layoutHelper: System seems unavailable", re);
-
                     runtime.systemDownHandler.run();
-                    runtime.invalidateLayout();
-                    Sleep.sleepUninterruptibly(retryRate);
                 } else {
                     throw re;
                 }
                 if (rethrowAllExceptions) {
                     throw new RuntimeException(re);
                 }
+                runtime.invalidateLayout();
+                Sleep.sleepUninterruptibly(retryRate);
```

Comment:
I have added this sleep here. This thread will now sleep for `retryRate` duration in case of WrongEpochException also.
I have added this as I ran into a case where this thread was invalidating the runtime but never yielded to the runtime thread to actually invalidate the layout.
However, this also means that now, the data path can see delays up to 5 seconds on epoch changes.
Let me know if this is the right solution to this.

Output snippet from my experiment:
```
2018-04-19 14:09:06,966 WARN  [Thread-2] o.c.r.v.AbstractView - Got a wrong epoch exception, updating epoch to 0 and invalidate view
2018-04-19 14:09:06,975 WARN  [Thread-2] o.c.r.v.AbstractView - Got a wrong epoch exception, updating epoch to 0 and invalidate view
2018-04-19 14:09:06,986 WARN  [Thread-2] o.c.r.v.AbstractView - Got a wrong epoch exception, updating epoch to 0 and invalidate view
...
2018-04-19 15:34:36,180 WARN  [Thread-2] o.c.r.v.AbstractView - Got a wrong epoch exception, updating epoch to 0 and invalidate view
2018-04-19 15:34:36,182 WARN  [Thread-2] o.c.r.v.AbstractView - Got a wrong epoch exception, updating epoch to 0 and invalidate view
2018-04-19 15:34:36,185 WARN  [Thread-2] o.c.r.v.AbstractView - Got a wrong epoch exception, updating epoch to 0 and invalidate view
```

Reply:
Why invalidating in case of ServerNotReady?

Reply:
Consider the following case:
The client is stuck at a stale epoch (by requesting a layout from a trailing minority.)
Now, this client accesses a stale sequencer for an epoch. This stale/secondary sequencer has been sealed and cannot issue tokens anymore and will throw the ServerNotReady Exception. If the client does not invalidate the layout, it will never realize that it has been trying to fetch tokens from the wrong sequencer the entire time.

Reply:
> This stale/secondary sequencer has been sealed and cannot issue tokens anymore and will throw the ServerNotReady Exception

I don't think that is true. If a client with an old layout tries to request tokens from a sealed sequencer, it would get a WrongEpochException, because validateEpoch is called before isServerReadyToHandleMsg

https://github.com/CorfuDB/CorfuDB/blob/master/infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java#L146